File: packages/compiler/src/generators/component.ts
import {
  getDelimitersArrayNewline,
  getDelimitersObject,
  getDelimitersObjectNewline,
} from '@vue-jsx-vapor/compiler-rs'
import { camelize, isArray } from '@vue/shared'
import {
  IRDynamicPropsKind,
  IRSlotType,
  type BlockIRNode,
  type CreateComponentIRNode,
  type IRProp,
  type IRProps,
  type IRPropsStatic,
  type IRSlotDynamic,
  type IRSlotDynamicBasic,
  type IRSlotDynamicConditional,
  type IRSlotDynamicLoop,
  type IRSlots,
  type IRSlotsStatic,
} from '../ir'
import {
  createSimpleExpression,
  genCall,
  genMulti,
  INDENT_END,
  INDENT_START,
  NEWLINE,
  toValidAssetId,
  walkIdentifiers,
  type CodeFragment,
} from '../utils'
import type { CodegenContext } from '../generate'
import { genBlock } from './block'
import { genDirectiveModifiers, genDirectivesForElement } from './directive'
import { genEventHandler } from './event'
import { genExpression } from './expression'
import { genPropKey, genPropValue } from './prop'
import { genModelHandler } from './vModel'

export function genCreateComponent(
  operation: CreateComponentIRNode,
  context: CodegenContext,
): CodeFragment[] {
  const { helper } = context

  const tag = genTag()
  const { root, props, slots, once } = operation
  const rawProps = genRawProps(props, context)
  const rawSlots = genRawSlots(slots, context)

  return [
    NEWLINE,
    `const n${operation.id} = `,
    ...genCall(
      operation.dynamic && !operation.dynamic.isStatic
        ? helper('createDynamicComponent')
        : operation.asset
          ? helper('createComponentWithFallback')
          : helper('createComponent'),
      [tag, rawProps, rawSlots, root ? 'true' : null, once ? 'true' : null],
    ),
    ...genDirectivesForElement(operation.id, context),
  ]

  function genTag() {
    if (operation.dynamic) {
      if (operation.dynamic.isStatic) {
        return genCall(helper('resolveDynamicComponent'), [
          genExpression(operation.dynamic, context),
        ])
      } else {
        return ['() => (', ...genExpression(operation.dynamic, context), ')']
      }
    } else if (operation.asset) {
      return toValidAssetId(operation.tag, 'component')
    } else {
      return genExpression(createSimpleExpression(operation.tag), context)
    }
  }
}

export function genRawProps(
  props: IRProps[],
  context: CodegenContext,
): CodeFragment[] | undefined {
  const staticProps = props[0]
  if (isArray(staticProps)) {
    if (!staticProps.length && props.length === 1) {
      return
    }
    return genStaticProps(
      staticProps,
      context,
      genDynamicProps(props.slice(1), context),
    )
  } else if (props.length) {
    // all dynamic
    return genStaticProps([], context, genDynamicProps(props, context))
  }
}

function genStaticProps(
  props: IRPropsStatic,
  context: CodegenContext,
  dynamicProps?: CodeFragment[],
): CodeFragment[] {
  const args = props.map((prop) => genProp(prop, context, true))
  if (dynamicProps) {
    args.push([`$: `, ...dynamicProps])
  }
  return genMulti(
    args.length > 1 ? getDelimitersObjectNewline() : getDelimitersObject(),
    args,
  )
}

function genDynamicProps(
  props: IRProps[],
  context: CodegenContext,
): CodeFragment[] | undefined {
  const { helper } = context
  const frags: CodeFragment[][] = []
  for (const p of props) {
    let expr: CodeFragment[]
    if (isArray(p)) {
      if (p.length) {
        frags.push(genStaticProps(p, context))
      }
      continue
    } else if (p.kind === IRDynamicPropsKind.ATTRIBUTE)
      expr = genMulti(getDelimitersObject(), [genProp(p, context)])
    else {
      expr = genExpression(p.value, context)
      if (p.handler) expr = genCall(helper('toHandlers'), [expr])
    }
    frags.push(['() => (', ...expr, ')'])
  }
  if (frags.length) {
    return genMulti(getDelimitersArrayNewline(), frags)
  }
}

function genProp(prop: IRProp, context: CodegenContext, isStatic?: boolean) {
  const values = genPropValue(prop.values, context)
  return [
    ...genPropKey(prop, context),
    ': ',
    ...(prop.handler
      ? genEventHandler(
          context,
          prop.values[0],
          prop.handlerModifiers,
          true /* wrap handlers passed to components */,
        )
      : isStatic
        ? ['() => (', ...values, ')']
        : values),
    ...(prop.model
      ? [...genModelEvent(prop, context), ...genModelModifiers(prop, context)]
      : []),
  ]
}

function genModelEvent(prop: IRProp, context: CodegenContext): CodeFragment[] {
  const name = prop.key.isStatic
    ? [JSON.stringify(`onUpdate:${camelize(prop.key.content)}`)]
    : ['["onUpdate:" + ', ...genExpression(prop.key, context), ']']
  const handler = genModelHandler(prop.values[0], context)

  return [',', NEWLINE, ...name, ': () => ', ...handler]
}

function genModelModifiers(
  prop: IRProp,
  context: CodegenContext,
): CodeFragment[] {
  const { key, modelModifiers } = prop
  if (!modelModifiers || !modelModifiers.length) return []

  const modifiersKey = key.isStatic
    ? [`${key.content}Modifiers`]
    : ['[', ...genExpression(key, context), ' + "Modifiers"]']

  const modifiersVal = genDirectiveModifiers(modelModifiers)
  return [',', NEWLINE, ...modifiersKey, `: () => ({ ${modifiersVal} })`]
}

function genRawSlots(slots: IRSlots[], context: CodegenContext) {
  if (!slots.length) return
  const staticSlots = slots[0]
  if (staticSlots.slotType === IRSlotType.STATIC) {
    // single static slot
    return genStaticSlots(
      staticSlots,
      context,
      slots.length > 1 ? slots.slice(1) : undefined,
    )
  } else {
    return genStaticSlots(
      { slotType: IRSlotType.STATIC, slots: {} },
      context,
      slots,
    )
  }
}

function genStaticSlots(
  { slots }: IRSlotsStatic,
  context: CodegenContext,
  dynamicSlots?: IRSlots[],
) {
  const args = Object.keys(slots).map((name) => [
    `${JSON.stringify(name)}: `,
    ...genSlotBlockWithProps(slots[name], context),
  ])
  if (dynamicSlots) {
    args.push([`$: `, ...genDynamicSlots(dynamicSlots, context)])
  }
  return genMulti(getDelimitersObjectNewline(), args)
}

function genDynamicSlots(
  slots: IRSlots[],
  context: CodegenContext,
): CodeFragment[] {
  return genMulti(
    getDelimitersArrayNewline(),
    slots.map((slot) =>
      slot.slotType === IRSlotType.STATIC
        ? genStaticSlots(slot, context)
        : slot.slotType === IRSlotType.EXPRESSION
          ? slot.slots.content
          : genDynamicSlot(slot, context, true),
    ),
  )
}

function genDynamicSlot(
  slot: IRSlotDynamic,
  context: CodegenContext,
  withFunction = false,
): CodeFragment[] {
  let frag: CodeFragment[]
  switch (slot.slotType) {
    case IRSlotType.DYNAMIC:
      frag = genBasicDynamicSlot(slot, context)
      break
    case IRSlotType.LOOP:
      frag = genLoopSlot(slot, context)
      break
    case IRSlotType.CONDITIONAL:
      frag = genConditionalSlot(slot, context)
      break
  }
  return withFunction ? ['() => (', ...frag, ')'] : frag
}

function genBasicDynamicSlot(
  slot: IRSlotDynamicBasic,
  context: CodegenContext,
): CodeFragment[] {
  const { name, fn } = slot
  return genMulti(getDelimitersObjectNewline(), [
    ['name: ', ...genExpression(name, context)],
    ['fn: ', ...genSlotBlockWithProps(fn, context)],
  ])
}

function genLoopSlot(
  slot: IRSlotDynamicLoop,
  context: CodegenContext,
): CodeFragment[] {
  const { name, fn, loop } = slot
  const { value, key, index, source } = loop
  const rawValue = value && value.content
  const rawKey = key && key.content
  const rawIndex = index && index.content

  const idMap: Record<string, string> = {}
  if (rawValue) idMap[rawValue] = rawValue
  if (rawKey) idMap[rawKey] = rawKey
  if (rawIndex) idMap[rawIndex] = rawIndex
  const slotExpr = genMulti(getDelimitersObjectNewline(), [
    ['name: ', ...context.withId(() => genExpression(name, context), idMap)],
    [
      'fn: ',
      ...context.withId(() => genSlotBlockWithProps(fn, context), idMap),
    ],
  ])
  return [
    ...genCall(context.helper('createForSlots'), [
      genExpression(source!, context),
      [
        ...genMulti(
          ['(', ')', ', ', undefined],
          [
            rawValue ? rawValue : rawKey || rawIndex ? '_' : null,
            rawKey ? rawKey : rawIndex ? '__' : null,
            rawIndex,
          ],
        ),
        ' => (',
        ...slotExpr,
        ')',
      ],
    ]),
  ]
}

function genConditionalSlot(
  slot: IRSlotDynamicConditional,
  context: CodegenContext,
): CodeFragment[] {
  const { condition, positive, negative } = slot
  return [
    ...genExpression(condition, context),
    INDENT_START,
    NEWLINE,
    '? ',
    ...genDynamicSlot(positive, context),
    NEWLINE,
    ': ',
    ...(negative ? [...genDynamicSlot(negative, context)] : ['void 0']),
    INDENT_END,
  ]
}

function genSlotBlockWithProps(oper: BlockIRNode, context: CodegenContext) {
  let isDestructureAssignment = false
  let rawProps: string | undefined
  let propsName: string | undefined
  let exitScope: (() => void) | undefined
  let depth: number | undefined
  const props = oper.props!
  const idsOfProps = new Set<string>()

  if (props) {
    rawProps = props.content
    if ((isDestructureAssignment = !!props.ast)) {
      ;[depth, exitScope] = context.enterScope()
      propsName = `_slotProps${depth}`
      walkIdentifiers(
        props.ast,
        (id, _, __, isReference, isLocal) => {
          if (isReference && !isLocal) idsOfProps.add(id.name)
        },
        true,
      )
    } else {
      idsOfProps.add((propsName = rawProps))
    }
  }

  const idMap: Record<string, string | null> = {}

  idsOfProps.forEach(
    (id) =>
      (idMap[id] = isDestructureAssignment
        ? `${propsName}[${JSON.stringify(id)}]`
        : null),
  )
  const blockFn = context.withId(
    () => genBlock(oper, context, [propsName]),
    idMap,
  )
  exitScope && exitScope()

  return blockFn
}

File: packages/compiler-rs/src/generate/component.rs
use std::mem;

use napi::bindgen_prelude::Either3;
use oxc_allocator::CloneIn;
use oxc_ast::NONE;
use oxc_ast::ast::BinaryOperator;
use oxc_ast::ast::BindingPatternKind;
use oxc_ast::ast::Expression;
use oxc_ast::ast::FormalParameterKind;
use oxc_ast::ast::ObjectPropertyKind;
use oxc_ast::ast::PropertyKind;
use oxc_ast::ast::Statement;
use oxc_ast::ast::VariableDeclarationKind;
use oxc_span::SPAN;

use crate::generate::CodegenContext;
use crate::generate::directive::gen_directive_modifiers;
use crate::generate::directive::gen_directives_for_element;
use crate::generate::event::gen_event_handler;
use crate::generate::expression::gen_expression;
use crate::generate::prop::gen_prop_key;
use crate::generate::prop::gen_prop_value;
use crate::generate::slot::gen_raw_slots;
use crate::generate::v_model::gen_model_handler;
use crate::ir::component::IRProp;
use crate::ir::component::IRProps;
use crate::ir::component::IRPropsStatic;
use crate::ir::index::BlockIRNode;
use crate::ir::index::CreateComponentIRNode;
use crate::ir::index::Modifiers;
use crate::ir::index::SimpleExpressionNode;
use crate::utils::text::camelize;
use crate::utils::text::to_valid_asset_id;

pub fn gen_create_component<'a>(
  statements: &mut oxc_allocator::Vec<'a, Statement<'a>>,
  operation: CreateComponentIRNode<'a>,
  context: &'a CodegenContext<'a>,
  context_block: &'a mut BlockIRNode<'a>,
) {
  let ast = &context.ast;
  let CreateComponentIRNode {
    tag,
    root,
    props,
    slots,
    once,
    id,
    dynamic,
    asset,
    ..
  } = operation;

  let is_dynamic = if let Some(dynamic) = &dynamic
    && !dynamic.is_static
  {
    true
  } else {
    false
  };
  let tag = if let Some(dynamic) = dynamic {
    if dynamic.is_static {
      ast
        .expression_call(
          SPAN,
          ast.expression_identifier(SPAN, ast.atom(&context.helper("resolveDynamicComponent"))),
          NONE,
          ast.vec1(gen_expression(dynamic, context, None, None).into()),
          false,
        )
        .into()
    } else {
      ast
        .expression_arrow_function(
          SPAN,
          true,
          false,
          NONE,
          ast.formal_parameters(
            SPAN,
            FormalParameterKind::ArrowFormalParameters,
            ast.vec(),
            NONE,
          ),
          NONE,
          ast.function_body(
            SPAN,
            ast.vec(),
            ast.vec1(ast.statement_expression(SPAN, gen_expression(dynamic, context, None, None))),
          ),
        )
        .into()
    }
  } else if asset {
    ast
      .expression_identifier(SPAN, ast.atom(&to_valid_asset_id(&tag, "component")))
      .into()
  } else {
    gen_expression(
      SimpleExpressionNode {
        content: tag,
        is_static: false,
        loc: SPAN,
        ast: None,
      },
      context,
      None,
      None,
    )
    .into()
  };

  let raw_props = gen_raw_props(props, context);
  let _context_block = context_block as *mut BlockIRNode;
  let raw_slots = gen_raw_slots(slots, context, unsafe { &mut *_context_block });

  let mut arguments = ast.vec1(tag);
  if let Some(raw_props) = raw_props {
    arguments.push(raw_props.into());
  } else if root || once || raw_slots.is_some() {
    arguments.push(ast.expression_null_literal(SPAN).into());
  }
  if let Some(raw_slots) = raw_slots {
    arguments.push(raw_slots.into());
  } else if root || once {
    arguments.push(ast.expression_null_literal(SPAN).into());
  }
  if root {
    arguments.push(ast.expression_boolean_literal(SPAN, true).into());
  } else if once {
    arguments.push(ast.expression_null_literal(SPAN).into())
  }
  if once {
    arguments.push(ast.expression_boolean_literal(SPAN, true).into());
  }
  statements.push(Statement::VariableDeclaration(
    ast.alloc_variable_declaration(
      SPAN,
      VariableDeclarationKind::Const,
      ast.vec1(ast.variable_declarator(
        SPAN,
        VariableDeclarationKind::Const,
        ast.binding_pattern(
          BindingPatternKind::BindingIdentifier(
            ast.alloc_binding_identifier(SPAN, ast.atom(&format!("n{id}"))),
          ),
          NONE,
          false,
        ),
        Some(ast.expression_call(
          SPAN,
          ast.expression_identifier(
            SPAN,
            ast.atom(&context.helper(if is_dynamic {
              "createDynamicComponent"
            } else if asset {
              "createComponentWithFallback"
            } else {
              "createComponent"
            })),
          ),
          NONE,
          arguments,
          false,
        )),
        false,
      )),
      false,
    ),
  ));
  if let Some(directive_statement) = gen_directives_for_element(id, context, context_block) {
    statements.push(directive_statement);
  }
}

fn gen_raw_props<'a>(
  mut props: Vec<IRProps<'a>>,
  context: &'a CodegenContext<'a>,
) -> Option<Expression<'a>> {
  let props_len = props.len();
  if let Either3::A(static_props) = &props[0] {
    if static_props.len() == 0 && props_len == 1 {
      return None;
    }
    let static_props = props.remove(0);
    if let Either3::A(static_props) = static_props {
      Some(gen_static_props(
        static_props,
        context,
        gen_dynamic_props(props, context),
      ))
    } else {
      None
    }
  } else if props_len > 0 {
    // all dynamic
    Some(gen_static_props(
      vec![],
      context,
      gen_dynamic_props(props, context),
    ))
  } else {
    None
  }
}

fn gen_static_props<'a>(
  props: IRPropsStatic<'a>,
  context: &'a CodegenContext<'a>,
  dynamic_props: Option<Expression<'a>>,
) -> Expression<'a> {
  let ast = &context.ast;
  let mut properties = ast.vec();
  let _properties = &mut properties as *mut oxc_allocator::Vec<ObjectPropertyKind>;
  for prop in props {
    gen_prop(unsafe { &mut *_properties }, prop, context, true)
  }
  if let Some(dynamic_props) = dynamic_props {
    properties.push(ast.object_property_kind_object_property(
      SPAN,
      PropertyKind::Init,
      ast.property_key_static_identifier(SPAN, ast.atom("$")),
      dynamic_props,
      false,
      false,
      false,
    ));
  }
  ast.expression_object(SPAN, properties)
}

fn gen_dynamic_props<'a>(
  props: Vec<IRProps<'a>>,
  context: &'a CodegenContext<'a>,
) -> Option<Expression<'a>> {
  let ast = &context.ast;
  let mut frags = ast.vec();
  for p in props {
    let mut expr = None;
    if let Either3::A(p) = p {
      if p.len() > 0 {
        frags.push(gen_static_props(p, context, None))
      }
      continue;
    } else if let Either3::B(p) = p {
      let mut properties = ast.vec();
      gen_prop(&mut properties, p, context, false);
      expr = Some(ast.expression_object(SPAN, properties));
    } else if let Either3::C(p) = p {
      let expression = gen_expression(p.value, context, None, None);
      expr = if p.handler.unwrap_or_default() {
        Some(ast.expression_call(
          SPAN,
          ast.expression_identifier(SPAN, ast.atom(&context.helper("toHandlers"))),
          NONE,
          ast.vec1(expression.into()),
          false,
        ))
      } else {
        Some(expression)
      }
    }
    frags.push(ast.expression_arrow_function(
      SPAN,
      true,
      false,
      NONE,
      ast.formal_parameters(
        SPAN,
        FormalParameterKind::ArrowFormalParameters,
        ast.vec(),
        NONE,
      ),
      NONE,
      ast.function_body(
        SPAN,
        ast.vec(),
        if let Some(expr) = expr {
          ast.vec1(ast.statement_expression(SPAN, expr))
        } else {
          ast.vec()
        },
      ),
    ));
  }
  if frags.len() > 0 {
    return Some(
      ast.expression_array(SPAN, ast.vec_from_iter(frags.into_iter().map(|i| i.into()))),
    );
  }
  None
}

fn gen_prop<'a>(
  properties: &mut oxc_allocator::Vec<'a, ObjectPropertyKind<'a>>,
  mut prop: IRProp<'a>,
  context: &'a CodegenContext<'a>,
  is_static: bool,
) {
  let ast = &context.ast;
  let model = prop.model.unwrap_or_default();
  let handler = prop.handler.unwrap_or_default();
  let Modifiers {
    keys,
    non_keys,
    options,
  } = prop.handler_modifiers.unwrap_or(Modifiers {
    keys: vec![],
    non_keys: vec![],
    options: vec![],
  });
  let mut values = mem::take(&mut prop.values);

  let model_modifiers = prop.model_modifiers.take();
  let model = if model {
    Some(gen_model(
      prop.key.clone(),
      values[0].clone(),
      model_modifiers,
      context,
    ))
  } else {
    None
  };

  let value = if handler {
    gen_event_handler(
      context,
      Some(values.remove(0)),
      keys,
      non_keys,
      true, /* wrap handlers passed to components */
    )
  } else {
    let values = gen_prop_value(values, context);
    if is_static {
      ast.expression_arrow_function(
        SPAN,
        true,
        false,
        NONE,
        ast.formal_parameters(
          SPAN,
          FormalParameterKind::ArrowFormalParameters,
          ast.vec(),
          NONE,
        ),
        NONE,
        ast.function_body(
          SPAN,
          ast.vec(),
          ast.vec1(ast.statement_expression(SPAN, values)),
        ),
      )
    } else {
      values
    }
  };

  let key = gen_prop_key(
    prop.key,
    prop.runtime_camelize,
    prop.modifier,
    handler,
    options,
    context,
  );
  let computed = key.is_expression();
  properties.push(ast.object_property_kind_object_property(
    SPAN,
    PropertyKind::Init,
    key,
    value,
    false,
    false,
    computed,
  ));

  if let Some(model) = model {
    properties.extend(model);
  }
}

fn gen_model<'a>(
  key: SimpleExpressionNode<'a>,
  value: SimpleExpressionNode<'a>,
  model_modifiers: Option<Vec<String>>,
  context: &'a CodegenContext<'a>,
) -> oxc_allocator::Vec<'a, ObjectPropertyKind<'a>> {
  let ast = &context.ast;
  let mut properties = ast.vec();
  let is_static = key.is_static;
  let content = key.content.clone();
  let expression = gen_expression(key, context, None, None);

  let modifiers = if let Some(model_modifiers) = model_modifiers
    && model_modifiers.len() > 0
  {
    let modifers_key = if is_static {
      ast
        .property_key_static_identifier(SPAN, ast.atom(&format!("{}Modifiers", camelize(&content))))
    } else {
      ast
        .expression_binary(
          SPAN,
          expression.clone_in(context.ast.allocator),
          BinaryOperator::Addition,
          ast.expression_string_literal(SPAN, ast.atom("Modifiers"), None),
        )
        .into()
    };
    let modifiers_val = Expression::ObjectExpression(gen_directive_modifiers(model_modifiers, ast));

    Some(ast.object_property_kind_object_property(
      SPAN,
      PropertyKind::Init,
      modifers_key,
      ast.expression_arrow_function(
        SPAN,
        true,
        false,
        NONE,
        ast.formal_parameters(
          SPAN,
          FormalParameterKind::ArrowFormalParameters,
          ast.vec(),
          NONE,
        ),
        NONE,
        ast.function_body(
          SPAN,
          ast.vec(),
          ast.vec1(ast.statement_expression(SPAN, modifiers_val)),
        ),
      ),
      false,
      false,
      !is_static,
    ))
  } else {
    None
  };

  let name = if is_static {
    ast.property_key_static_identifier(
      SPAN,
      ast.atom(&format!("\"onUpdate:{}\"", camelize(&content))),
    )
  } else {
    ast
      .expression_binary(
        SPAN,
        ast.expression_string_literal(SPAN, ast.atom("onUpdate:"), None),
        BinaryOperator::Addition,
        expression,
      )
      .into()
  };

  let handler = gen_model_handler(value, context);
  properties.push(ast.object_property_kind_object_property(
    SPAN,
    PropertyKind::Init,
    name,
    ast.expression_arrow_function(
      SPAN,
      true,
      false,
      NONE,
      ast.formal_parameters(
        SPAN,
        FormalParameterKind::ArrowFormalParameters,
        ast.vec(),
        NONE,
      ),
      NONE,
      ast.function_body(
        SPAN,
        ast.vec(),
        ast.vec1(ast.statement_expression(SPAN, handler)),
      ),
    ),
    false,
    false,
    !is_static,
  ));

  if let Some(modifiers) = modifiers {
    properties.push(modifiers)
  }
  properties
}
